### PR TITLE
Add customizable local storage handlers

### PR DESCRIPTION
**Pull request by jplafonta**
* Add signatures for Local Storage handlers and new public API to configure them.  Local Storage handlers use XAsync interface (client will need to implement XAsyncProvider for those operations)
* Added internal LocalStorage class which wraps invocations of those client hooks and serializes local storage operations.  This means that client provided hooks don't have to worry about synchronizing simultaneous operations on the same file.
* Added some helper classes that reduce duplicated code needed for calling XAsync APIs.  They should eventually be used for LHC API invocations as well.
* Made some improvements to ensure file tracing completes before cleanup returns
* Minor refactoring

### CODE REVIEW

File: source/code/source/Platform/LocalStorageHandlers.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+) noexcept;
+
+HRESULT STDAPIVCALLTYPE DefaultLocalStorageClearAsync(
+    _In_opt_ void* /*context*/,
+    _In_z_ const char* key,
+    _Inout_ XAsyncBlock* async
+);
+
+}
+}
```
Comment (jplafonta) [inline source/code/source/Platform/LocalStorageHandlers.h:29]: newline

File: source/code/source/Platform/LocalStorageHandlers.cpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        Vector<uint8_t>{}
+    } };
+
+    RETURN_IF_FAILED(XAsyncBegin(async, providerContext.get(), nullptr, __FUNCTION__, LocalStorageOperationAsyncProvider));
+    providerContext.release();
+    return S_OK;
+}
+
+}
+}
```
Comment (jplafonta) [inline source/code/source/Platform/LocalStorageHandlers.cpp:218]: newline

File: source/code/source/Common/XAsyncOperation.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    XAsyncOperationBase& operator=(XAsyncOperationBase const&) noexcept = delete;
+    virtual ~XAsyncOperationBase() noexcept = default;
+
+private: // Interface to be implemented by XAsync wrapper Operations
+    virtual HRESULT OnStarted(XAsyncBlock* async) = 0;
+    virtual void OnFailed(HRESULT hr) = 0;
+    virtual void OnCompleted(XAsyncBlock* async) = 0;
+
+private: // IOperation implementation
+    void OnStarted(UniquePtr<IOperation> self) noexcept override;
```
Comment (jplafonta) [inline source/code/source/Common/XAsyncOperation.h:46]: > void OnStarted(UniquePtr<IOperation> self) noexcept override; [](http://example.com/codeflow?start=3&length=62)

final

**Review by jasonsandlin — APPROVED** [review]

File: source/code/source/Common/XAsyncOperation.h (hunk 3 above)
Comment (jplafonta) [reply source/code/source/Common/XAsyncOperation.h:46]: (Will be addressed in next PR)

**Pull request merged**